jsonml.go

```diff
@@ -0,0 +1,145 @@
+/*
+Copyright 2019 Google Inc. All rights reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package jsonnet
+
+import (
+	"encoding/xml"
+	"errors"
+	"fmt"
+	"io"
+	"strings"
+)
+
+type stack struct {
+	elements []interface{}
+}
+
+func (s *stack) Push(e interface{}) {
+	s.elements = append(s.elements, e)
+}
+
+func (s *stack) Pop() (interface{}, error) {
+	if len(s.elements) == 0 {
+		return nil, errors.New("cannot pop from empty stack")
+	}
+	l := len(s.elements)
+	e := s.elements[l-1]
+	s.elements = s.elements[:l-1]
+	return e, nil
+}
+
+func (s *stack) Size() int {
+	return len(s.elements)
+}
+
+type jsonMLBuilder struct {
+	stack     *stack
+	currDepth int
+}
+
+// BuildJsonmlFromString returns a jsomML form of given xml string.
+func BuildJsonmlFromString(s string) ([]interface{}, error) {
+	b := newBuilder()
+	d := xml.NewDecoder(strings.NewReader(s))
+
+	for {
+		token, err := d.Token()
+		if err == io.EOF {
+			break
+		}
+
+		if err != nil {
+			return nil, err
+		}
+
+		if err := b.addToken(token); err != nil {
+			return nil, err
+		}
+	}
+
+	if b.stack.Size() == 0 {
+		// No nodes has been identified
+		return nil, fmt.Errorf("%s is not a valid XML", s)
+	}
+
+	return b.build(), nil
+}
+
+func newBuilder() *jsonMLBuilder {
+	return &jsonMLBuilder{
+		stack: &stack{},
+	}
+}
+
+func (b *jsonMLBuilder) addToken(token xml.Token) error {
+	switch token.(type) {
+	case xml.StartElement:
+		// check for multiple roots
+		if b.currDepth == 0 && b.stack.Size() > 0 {
+			// There are multiple root elements
+			return errors.New("XML cannot have multiple root elements")
+		}
+
+		t := token.(xml.StartElement)
+		node := []interface{}{t.Name.Local}
+		// Add Attributes
+		if len(t.Attr) > 0 {
+			attr := make(map[string]interface{})
+			for _, a := range t.Attr {
+				attr[a.Name.Local] = a.Value
+			}
+			node = append(node, attr)
+		}
+		b.stack.Push(node)
+		b.currDepth++
+	case xml.CharData:
+		t := token.(xml.CharData)
+		s := strings.TrimSpace(string(t))
```

Comment:
Is there a reason for trimming it?  I think whitespace is actually meaningful in XML because of things like `<b>foo</b>bar` vs `<b>foo</b> bar`.  If we're only supporting the (admittedly common) case of the XML file being whitespace agnostic (i.e. whoever consumes the XML doesn't care about the whitespace) then we can document it as not supporting that.  However I think just not trimming the strings here will fix it, will it not?

Reply:
The reason behind trimming whitespaces is to simplify jsonml. Currently due to formatted xml, jsonml becomes bit complex to work with due to empty content. We want to ignore empty lines but I agree current implementation would also trim spaces around the content. I could take of it.

I am thinking of allowing users to control this behaviour. We can provide an argument `trimWhitespaces = true` to the function which would control if user wants to remove formatting whitespaces. 

WDYT?

Reply:
I found this which looks interesting.  http://xml.coverpages.org/rfc-wshp19990416.html  I wonder if the current behaviour of the manifester is wrong?

Reply:
What I understand from these link is it is safe to remove boundary whitespaces as a default behaviour. We would provide an argument `preserveWhitespaces = false` to preserve them if anyone wants to do so.

Reply:
There seems to be some diversity in what other parsers are doing so I think anything we do is fine as long as it's documented.

Reply:
Updated